test/extensions/transport_sockets/starttls/upstream_starttls_integration_test.cc

```diff
@@ -0,0 +1,409 @@
+
+        auto startTlsTransportSocket =
+            Extensions::TransportSockets::StartTls::StartTlsSocketFactory(move(cleartext_context_),
+                                                                          move(server_tls_context_))
+                .createTransportSocket(std::make_shared<Network::TransportSocketOptionsImpl>(
+                    absl::string_view(""), std::vector<std::string>(), std::vector<std::string>()));
+
+        server_connection_ = dispatcher_->createServerConnection(
+            std::move(socket), move(startTlsTransportSocket), stream_info_);
+        server_connection_->addReadFilter(std::make_shared<TerminalServerTlsFilter>());
+      }));
+
+  listener_ =
+      dispatcher_->createListener(socket, listener_callbacks_, true, ENVOY_TCP_BACKLOG_SIZE);
+
+  // Add a start_tls cluster
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    *bootstrap.mutable_static_resources()->add_clusters() = ConfigHelper::buildStartTlsCluster(
+        socket->addressProvider().localAddress()->ip()->addressAsString(),
+        socket->addressProvider().localAddress()->ip()->port());
+  });
+
+  BaseIntegrationTest::initialize();
+
+  Network::Address::InstanceConstSharedPtr address =
+      Ssl::getSslAddress(version_, lookupPort("tcp_proxy"));
+
+  conn_ = std::make_unique<ClientTestConnection>(
+      *dispatcher_, address, Network::Address::InstanceConstSharedPtr(),
+      cleartext_context_->createTransportSocket(
+          std::make_shared<Network::TransportSocketOptionsImpl>(
+              absl::string_view(""), std::vector<std::string>(), std::vector<std::string>())),
+      nullptr);
+
+  conn_->enableHalfClose(true);
+  conn_->addConnectionCallbacks(connect_callbacks_);
+  conn_->addReadFilter(payload_reader_);
+}
+
+// Method adds StartTlsSwitchFilter into the filter chain.
+// The filter is required to instruct StartTls transport
+// socket to start using Tls.
+void StartTlsIntegrationTest::addStartTlsSwitchFilter(ConfigHelper& config_helper) {
+  config_helper.addNetworkFilter(R"EOF(
+      name: startTls
+      typed_config:
+        "@type": type.googleapis.com/test.integration.starttls.StartTlsFilterConfig
+    )EOF");
+  // double-check the filter was actually added
+  config_helper.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    ASSERT_EQ("startTls",
+              bootstrap.static_resources().listeners(0).filter_chains(0).filters(0).name());
+  });
+}
+
+// Test creates a clear-text connection from a client to Envoy and sends several messages.
+// Then a special message is sent, which causes StartTlsSwitchFilter to
+// instruct StartTls transport socket to start using tls.
+// The Client connection starts using tls, performs tls handshake and few messages
+// are sent over tls. start-tls transport socket de-crypts the messages and forwards them
+// upstream in clear-text..
+TEST_P(StartTlsIntegrationTest, SwitchToTlsFromClient) {
```

Comment:
I may be missing something, but how does this test verifies that UpstreamStartTls transport sockert switches to TLS?

Reply:
I'll update the doc string since it's not trivial to see. It happens in the `StartTlsSwitchFilter::upstreamWrite` method where we call `upstream_connection_->startSecureTransport()`. Currently line 149.